spaghetto/src/Spago/Command/Ls.purs

```diff
@@ -1,3 +1,120 @@
+import Spago.Prelude
+
+import Data.Array (replicate)
+import Data.Codec.Argonaut as CA
+import Data.Codec.Argonaut.Record as CAR
+import Data.Foldable (elem, maximum)
+import Data.Map (filterKeys)
+import Data.Map as Map
+import Data.Set (unions)
+import Data.String.CodeUnits (fromCharArray, length)
+import Data.Tuple.Nested (type (/\))
+import Registry.PackageName as PackageName
+import Registry.Version as Version
+import Spago.Bin.Flags (IncludeTransitive(..))
+import Spago.Config (Dependencies(..), Package(..), PackageSet, Workspace, getPackageLocation)
+import Spago.Config as Config
+import Spago.Json (stringifyJson)
+
+data JsonFlag = JsonOutputNo | JsonOutputYes
+
+type LsEnv =
+  { dependencies :: PackageSet
+  , logOptions :: LogOptions
+  , workspace :: Workspace
+  }
+
+type JsonPackageOutput =
+  { packageName :: String
+  -- TODO: Was Json Value, has tag
+  -- See https://github.com/purescript/spago/blob/ec22308d181340382494e5bd6ef9102a3c87c0bb/src/Spago/Types.hs#L38-L47
+  , repo :: String
+  , version :: String
+  }
+
+jsonPackageOutputCodec :: JsonCodec JsonPackageOutput
+jsonPackageOutputCodec = CAR.object "JsonPackageOutput"
+  { packageName: CA.string
+  , repo: CA.string
+  , version: CA.string
+  }
+
+listPackageSet :: JsonFlag -> Spago LsEnv Unit
+listPackageSet jsonFlag = do
+  logDebug "Running `listPackageSet`"
+  { workspace } <- ask
+  output $ formatPackageNames jsonFlag (Map.toUnfoldable workspace.packageSet)
+
+listPackages :: IncludeTransitive -> JsonFlag -> Spago LsEnv Unit
+listPackages packagesFilter jsonFlag = do
+  logDebug "Running `listPackages`"
+  { dependencies, workspace } <- ask
+  let
+    packagesToList = Map.toUnfoldable case packagesFilter of
+      IncludeTransitive -> dependencies
+      NoIncludeTransitive ->
+        let
+          workspacePackages = Config.getWorkspacePackages workspace.packageSet
+          toDependencyNames (Dependencies deps) = Map.keys deps
+          directDependencies = unions $ ((toDependencyNames <<< _.dependencies <<< _.package) <$> workspacePackages)
+        in
+          filterKeys (_ `elem` directDependencies) dependencies
+  case packagesToList of
+    [] -> logWarn "There are no dependencies listed in your spago.dhall"
+    _ -> output $ formatPackageNames jsonFlag packagesToList
+
+formatPackageNames :: forall a. JsonFlag -> Array (PackageName /\ Package) -> OutputFormat a
+formatPackageNames = case _ of
+  JsonOutputYes -> OutputLines <<< formatPackageNamesJson
+  JsonOutputNo -> OutputLines <<< formatPackageNamesText
+  where
+  formatPackageNamesJson :: Array (PackageName /\ Package) -> Array String
+  formatPackageNamesJson pkgs = (stringifyJson jsonPackageOutputCodec <<< asJson <$> pkgs)
+    where
+    asJson (name /\ package) =
+      { packageName: PackageName.print name
+      -- TODO: Do we want to stay backward compatible and have a tagged remote and local type?
+      -- Or do we create a codec for package (with 4 variants)?
+      , repo: "TODO"
```

Comment:
:question: In general I'm asking myself, if we should:
* map to `Remote | Local` like it was before (https://github.com/andys8/spago/blob/44793b36ffdb38d1bf43b4ca77a7d4ab513e9632/src/Spago/Types.hs#L27-L35)
* map to `Package`, and introduce a json coded (https://github.com/andys8/spago/blob/44793b36ffdb38d1bf43b4ca77a7d4ab513e9632/spaghetto/src/Spago/Config.purs#L293-L297)

Reply:
Keeping backwards compatibility is not a priority here (since there are so many breaking changes anyways), and since the new types are so different from the old ones I would say it's better to come up with something else rather than try to fit in the old format.

So let's discard the previous `Local`/`Remote`, and make a codec for `Package`

Reply:
Went with the coded: https://github.com/andys8/spago/blob/415149d6958a929a22a95ea7b85a1d1545e4b04a/spaghetto/src/Spago/Config.purs#L318-L332

Only part missing / not sure if necessary / how to handle, is `YamlDoc Config` in `WorkspacePackage`. I feel like printing this in dependencies isn't what we would want.

https://github.com/andys8/spago/blob/415149d6958a929a22a95ea7b85a1d1545e4b04a/spaghetto/src/Spago/Config.purs#L303-L310

Reply:
## Example JSON Output

```json
{
  "aff": "7.1.0",
  "aff-promise": "4.0.0",
  "affjax": "13.0.0",
  "affjax-node": "1.0.0",
  "argonaut-core": "7.0.0",
  "argparse-basic": "2.0.0",
  "arrays": "7.2.0",
  "bifunctors": "6.0.0",
  "codec-argonaut": "10.0.0",
  "console": "6.0.0",
  "control": "6.0.0",
  "datetime": "6.1.0",
  "dodo-printer": "2.2.1",
  "effect": "4.0.0",
  "either": "6.1.0",
  "exceptions": "6.0.0",
  "filterable": "5.0.0",
  "foldable-traversable": "6.0.0",
  "functions": "6.0.0",
  "http-methods": "6.0.0",
  "identity": "6.0.0",
  "integers": "6.0.0",
  "lists": "7.0.0",
  "maybe": "6.0.0",
  "newtype": "5.0.0",
  "node-buffer": "8.0.0",
  "node-execa": "1.1.0",
  "node-fs": "8.1.1",
  "node-fs-aff": "9.1.0",
  "node-path": "5.0.0",
  "node-process": "10.0.0",
  "now": "6.0.0",
  "nullable": "6.0.0",
  "ordered-collections": "3.0.0",
  "parallel": "6.0.0",
  "partial": "4.0.0",
  "prelude": "6.0.1",
  "profunctor": "6.0.0",
  "record": "4.0.0",
  "refs": "6.0.0",
  "registry-foreign": {
    "git": "https://github.com/purescript/registry-dev.git",
    "ref": "8113deecae25410ba62c31a5842c6ebbf1c12461",
    "subdir": "foreign"
  },
  "registry-lib": {
    "git": "https://github.com/purescript/registry-dev.git",
    "ref": "8113deecae25410ba62c31a5842c6ebbf1c12461",
    "subdir": "lib"
  },
  "spago": {
    "doc": null,
    "hasTests": true,
    "package": {
      "dependencies": [
        "aff",
        "aff-promise",
        "affjax",
        "affjax-node",
        "argonaut-core",
        "arrays",
        "bifunctors",
        "codec-argonaut",
        "console",
        "control",
        "datetime",
        "dodo-printer",
        "effect",
        "either",
        "exceptions",
        "filterable",
        "foldable-traversable",
        "functions",
        "http-methods",
        "identity",
        "integers",
        "lists",
        "maybe",
        "newtype",
        "node-buffer",
        "node-execa",
        "node-fs",
        "node-fs-aff",
        "node-path",
        "node-process",
        "now",
        "nullable",
        "ordered-collections",
        "parallel",
        "partial",
        "prelude",
        "profunctor",
        "record",
        "refs",
        "registry-foreign",
        "registry-lib",
        "strings",
        "transformers",
        "tuples",
        "unsafe-coerce"
      ],
      "name": "spago"
    },
    "path": "./"
  },
  "strings": "6.0.1",
  "transformers": "6.0.0",
  "tuples": "7.0.0",
  "unsafe-coerce": "6.0.0"
}
```